### ISSUE TITLE
    Rank-reducing linalg via reshape patterns + iree-hal-benchmark-dispatch-repeat-count seem to interact poorly and miscompile

### ISSUE THREAD

**Issue by nicolasvasilache** [iree-org/iree#12825]
### What happened?

I have been iterating on graph-level transformations and added this [repro](https://github.com/iree-org/iree-samples/commit/2dcd855116a4dab4b0cbcfb44fff4165cf249fa8) to iree-samples.

This is dependent on https://github.com/openxla/iree/pull/12822

Running     
    ```
    iree-compile --iree-hal-target-backends=cuda --iree-hal-benchmark-dispatch-repeat-count=5 transform_dialect/graph/cuda/bugs/2-ir.mlir | \
      iree-run-module --function=forward --device=cuda --input="3x17x768xf32=1"
    ```

Produces a different result when using `--iree-hal-benchmark-dispatch-repeat-count=5`.

I suspect this is not expected given that a difference in behavior is triggered by this [line](https://github.com/iree-org/iree-samples/commit/2dcd855116a4dab4b0cbcfb44fff4165cf249fa8#diff-fd396134be2ebc3284763e181fb313ddf86a2cea4087347b075171790c073e5eR58).

It seems to either be:
  1. an issue with one of the rank-reducing patterns via reshape or 
  2. a bufferization issue and we bufferize something inplace when we shouldn't (but it may be at the graph level ?)
  3. triggering a miscompile somewhere in IREE

**Comment by MaheshRavishankar** [comment]
This seems like a longer term issue... maybe better to look at this after #12824

**Comment by aaron-schneider** [comment]
Should we lower priority to P1 if we're going to wait?

**Comment by MaheshRavishankar** [comment]
I track P0 items as ones I need to look at in short order.

**Comment by aaron-schneider** [comment]
ping on p0 bug?

**Comment by MaheshRavishankar** [comment]
I guess these shouldnt be P0. Ill drop it down to P1.
